pkg/util/tracing: introduce `trace.span.force_verbose_regexp `

abarganier opened the pull request:
This patch introduces the `trace.verbose.span.operation.regexp` cluster
setting, which provides a surface for CRL employees debugging CRDB to
force certain tracing span operations to record in verbose mode via a
regular expression string. If the regexp matches a span's operation
name, that span will be forced into a verbose recording mode, capturing
all structured logs (recorded via `RecordStructured`) and events (Recorded
via `Recordf`) during the lifetime of the span.

We intentionally avoid release notes, as this setting is only meant
for internal usage.

Release note: none

Epic: CRDB-35278

Addresses: https://github.com/cockroachdb/cockroach/issues/117624

abarganier committed:
pkg/util/tracing: add regexp to force span recording mode to VERBOSE

This patch adds a mechanism to the tracer that allows us to specify
which span operations should be forced into a VERBOSE recording mode.

The mechanism works by adding a regular expression to the tracer, which
is used during span creation. If the span's operation name matches, then
the span is forced into a VERBOSE recording mode. This takes precedence
over any provided SpanOptions.

The result is essentially the same as if the span was created with the
two options `WithForceRealSpan()` and
`RecordingMode(tracingpb.RecordingVerbose)`.

The resulting span, and all its children, will record any structured
events logged via `RecordStructured()`, as well as events logged via
`Recordf()`.

Future patches will wire this mechanism up to a configurable cluster
setting, which will be private & only intended for use by L2 engineers.

Release note: none

abarganier committed:
pkg/util/tracing: introduce `trace.span.force_verbose_regexp`

This patch introduces the `trace.span.force_verbose_regexp` cluster
setting, which provides a surface for CRL employees debugging CRDB to
force certain tracing span operations to record in verbose mode via a
regular expression string. If the regexp matches a span's operation
name, that span will be forced into a verbose recording mode, capturing
all structured logs and events during the lifetime of the span.

We intentionally avoid release notes, as this setting is only meant
for internal usage.

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/118984)
<!-- Reviewable:end -->

abarganier commented:
Note: I'm going to follow up with some benchmarking, but wanted to get the implementation out asap to get feedback.

If benchmark results aren't good, we'll consider switching to a set of string operation names instead of a regular expression.

dhartunian reviewed (commented):
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/118984)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @abarganier and @dt)

___
*[`pkg/util/tracing/tracer.go` line 1214 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/118984#-NqA8qvdDSSLRNDWZTQO:-NqA8qve2TwNo1csVySh:b-16e4vc) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b352d68ff8c42c838cb1adf899f4882a5a34d28/pkg/util/tracing/tracer.go#L1214)):*
> ```Go
> 
> 	var recType tracingpb.RecordingType
> 	if forceVerbose || (!forceVerboseChecked && t.forceOpNameVerbose(opName)) {
> ```

can you use `shouldBeNoopSpan` instead of `!forceVerboseChecked`? That might clarify what's happening here. Then maybe you can avoid that bool entirely.


<!-- Sent from Reviewable.io -->

abarganier committed:
pkg/util/tracing: Add benchmark for `StartSpan()`

Upcoming tracing work will introduce a method for forcing spans
to record in verbose mode if their operation name matches a
regular expression provided via a cluster setting.

Tracing is a performance-sensitive area within CRDB, so we should
be aware of the performance impact of such changes.

This patch adds a simple benchmark for `StartSpan()`, so that we can
compare results after the changes have been made.

Release note: none

abarganier committed:
pkg/util/tracing: add benchmark for forceOpVerbose regex

Previous patches added the ability for operators to force certain spans
to record in verbose mode by setting an operation name regular
expression in the tracer via a cluster setting.

With these changes made, and the performance sensitivity of the tracing
library, we need to benchmark the impact of these changes.

This patch adds a new benchmark, BenchmarkStartSpan_OpNameRegexp, which
benchmarks the `StartSpan()` function when a regular expression is set.
The benchmark uses some operation names that match the regex, and others
that don't.

Additionally, we can run the previously added benchmark,
BenchmarkStartSpan, to measure the performance impact when a regular
expression is *not* set to see the impact when the new cluster setting
is not being used.

The results are below:

```
// Before changes
BenchmarkStartSpan
BenchmarkStartSpan-10    	 2565244	       476.5 ns/op	    1075 B/op	       2 allocs/op

// After changes
BenchmarkStartSpan
BenchmarkStartSpan-10    	 2490318	       483.2 ns/op	    1075 B/op	       2 allocs/op
```

```
// After changes, using regex
BenchmarkStartSpan_OpNameRegexp
BenchmarkStartSpan_OpNameRegexp-10    	 2082237	       574.0 ns/op	    1078 B/op	       2 allocs/op
```

In the case where the cluster setting is not being used, this represents
a mere 1.4% increase in execution time, which is acceptable.

If an operator uses the cluster setting, we see an 18.8% increase in
execution time when creating spans. Recording spans in verbose mode is
already well understood to hurt performance, so this is an acceptable
performance hit. If someone is using this cluster setting, then they are
already in debug mode and likely willing to degrade performance
temporarily in the interest of gaining additional debug information.

Release note: none

abarganier reviewed (commented):
TFTR, RFAL @dhartunian. 

I went ahead and added some benchmarks for span creation. I think the key thing to focus on here is performance when the new cluster setting *is not in use*, as this is the "happy path" that will impact day-to-day operations. 

I added a benchmark to measure this:
```
// Before changes
BenchmarkStartSpan
BenchmarkStartSpan-10            2565244               476.5 ns/op          1075 B/op          2 allocs/op

// After changes
BenchmarkStartSpan
BenchmarkStartSpan-10            2490318               483.2 ns/op          1075 B/op          2 allocs/op
```

That becomes our baseline, after which we can benchmark span creation when the cluster setting is enabled and the opName matches the regex.

```
// After changes, using regex
BenchmarkStartSpan_OpNameRegexp
BenchmarkStartSpan_OpNameRegexp-10       2082237               574.0 ns/op          1078 B/op          2 allocs/op
```

In the case where the cluster setting is not being used, this represents
a mere 1.4% increase in execution time, which is acceptable.

If an operator uses the cluster setting, we see an 18.8% increase in
execution time when creating spans. Recording spans in verbose mode is
already well understood to hurt performance, so this is an acceptable
performance hit. If someone is using this cluster setting, then they are
already in debug mode and likely willing to degrade performance
temporarily in the interest of gaining additional debug information.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/118984#-:-Nqn0LWJ4aaDdpmGHTtB:bdjqw3o)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @dt)

___
*[`pkg/util/tracing/tracer.go` line 1214 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/118984#-NqA8qvdDSSLRNDWZTQO:-Nqmn5_SDhn7GZsDbPbM:b-g08857) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b352d68ff8c42c838cb1adf899f4882a5a34d28/pkg/util/tracing/tracer.go#L1214)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

can you use `shouldBeNoopSpan` instead of `!forceVerboseChecked`? That might clarify what's happening here. Then maybe you can avoid that bool entirely.
</blockquote></details>

Yeah, these booleans are a mess lol. Regardless of the value of `shouldBeNoopSpan`, we have to call `forceOpNameVerbose()` at some point anyway. So might as well just make the check when we're doing the `shouldBeNoopSpan` and have a single additional boolean: `forceVerbose`. 


<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
> Reviewed 2 of 2 files at r3, 2 of 2 files at r4, 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/118984)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @abarganier and @dt)

___
*[`pkg/util/tracing/tracer.go` line 1214 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/118984#-NqA8qvdDSSLRNDWZTQO:-NrIDPaM-JDBJaW1qzB1:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b352d68ff8c42c838cb1adf899f4882a5a34d28/pkg/util/tracing/tracer.go#L1214)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

Yeah, these booleans are a mess lol. Regardless of the value of `shouldBeNoopSpan`, we have to call `forceOpNameVerbose()` at some point anyway. So might as well just make the check when we're doing the `shouldBeNoopSpan` and have a single additional boolean: `forceVerbose`.
</blockquote></details>

:thumbsup: 
___
*[`pkg/util/tracing/tracer.go` line 160 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/118984#-NrIH-4c2FUrKW4O5INV:-NrIH-4c2FUrKW4O5INW:bkh1jza) ([raw file](https://github.com/cockroachdb/cockroach/blob/03d89cdd153217218d503ec212ce5f842ae98c36/pkg/util/tracing/tracer.go#L160)):*
> ```Go
> var forceVerboseSpanRegexp = settings.RegisterStringSetting(
> 	settings.ApplicationLevel,
> 	"trace.verbose.span.operation.regexp",
> ```

I think the naming should have fewer dots. I think of the names as a hierarchy not as a "phrase" if that makes sense. In the `trace.*` group of cluster settings we don't have much structure right now (there's nothing else with `trace.verbose.*` prefix, or `trace.span.*` prefix for instance) so I think this can stay on the top level as something like `trace.span_force_verbose.regexp` which also matches your code naming.

If you like, `trace.span.force_verbose_regexp` also works since it's possible we might want to put stuff under `trace.span.*` in the future.


<!-- Sent from Reviewable.io -->

abarganier reviewed (commented):
RFAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/118984#-:-NrLhWSh5aZEein7fKaz:b-828fna)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @dt)

___
*[`pkg/util/tracing/tracer.go` line 1214 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/118984#-NqA8qvdDSSLRNDWZTQO:-NrLhVvX0g1-M5ggUxew:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b352d68ff8c42c838cb1adf899f4882a5a34d28/pkg/util/tracing/tracer.go#L1214)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

:thumbsup:
</blockquote></details>

Done.
___
*[`pkg/util/tracing/tracer.go` line 160 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/118984#-NrIH-4c2FUrKW4O5INV:-NrLhTnXDdf9diJuczMV:b-dfn4rf) ([raw file](https://github.com/cockroachdb/cockroach/blob/03d89cdd153217218d503ec212ce5f842ae98c36/pkg/util/tracing/tracer.go#L160)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

I think the naming should have fewer dots. I think of the names as a hierarchy not as a "phrase" if that makes sense. In the `trace.*` group of cluster settings we don't have much structure right now (there's nothing else with `trace.verbose.*` prefix, or `trace.span.*` prefix for instance) so I think this can stay on the top level as something like `trace.span_force_verbose.regexp` which also matches your code naming.

If you like, `trace.span.force_verbose_regexp` also works since it's possible we might want to put stuff under `trace.span.*` in the future.
</blockquote></details>

Good idea! I agree that's a much better name. Done.


<!-- Sent from Reviewable.io -->

dhartunian reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 2 of 2 files at r7, 2 of 2 files at r8, 1 of 1 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/118984#-:-Nr_zPHyE2lYxFm7I5sA:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @dt)



<!-- Sent from Reviewable.io -->

abarganier commented:
TFTR! 

bors r=dhartunian
